File: lang/java/cxlib/src/main/java/cx/BinaryDecoder.java
package cx;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * Decoder for the compact binary wire format produced by cx_to_ast_bin and
 * cx_to_events_bin.
 *
 * All integers are little-endian.
 *
 * String encoding:   u32(byte_len) + raw UTF-8 bytes (no null terminator)
 * OptStr encoding:   u8(0=absent | 1=present) + str if present
 * Attr encoding:     str:name + str:value_str + str:inferred_type
 */
public class BinaryDecoder {

    // ── Buffer reader ─────────────────────────────────────────────────────────

    static final class BufReader {
        private final ByteBuffer buf;

        BufReader(byte[] data) {
            this.buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN);
        }

        int u8() {
            return buf.get() & 0xFF;
        }

        int u16() {
            return buf.getShort() & 0xFFFF;
        }

        long u32() {
            return buf.getInt() & 0xFFFFFFFFL;
        }

        String str() {
            int len = (int) u32();
            byte[] bytes = new byte[len];
            buf.get(bytes);
            return new String(bytes, StandardCharsets.UTF_8);
        }

        String optStr() {
            int flag = u8();
            if (flag == 0) return null;
            return str();
        }
    }

    // ── Scalar coercion ───────────────────────────────────────────────────────

    static Object coerce(String inferredType, String valueStr) {
        return switch (inferredType) {
            case "int"    -> Long.parseLong(valueStr);
            case "float"  -> Double.parseDouble(valueStr);
            case "bool"   -> "true".equals(valueStr);
            case "null"   -> null;
            default       -> valueStr;  // string / date / datetime / bytes
        };
    }

    // ── Attr reader ───────────────────────────────────────────────────────────

    static Attr readAttr(BufReader b) {
        String name      = b.str();
        String valueStr  = b.str();
        String inferType = b.str();
        Object value     = coerce(inferType, valueStr);
        // Only carry dataType forward when it's not plain "string"
        String dataType  = "string".equals(inferType) ? null : inferType;
        return new Attr(name, value, dataType);
    }

    // ── AST node reader ───────────────────────────────────────────────────────

    static Node readNode(BufReader b) {
        int tid = b.u8();
        return switch (tid) {
            case 0x01 -> {
                String     name     = b.str();
                String     anchor   = b.optStr();
                String     dataType = b.optStr();
                String     merge    = b.optStr();
                int        nAttrs   = b.u16();
                List<Attr> attrs = new ArrayList<>(nAttrs);
                for (int i = 0; i < nAttrs; i++) attrs.add(readAttr(b));
                int        nItems   = b.u16();
                List<Node> items    = new ArrayList<>(nItems);
                for (int i = 0; i < nItems; i++) items.add(readNode(b));
                Element e = new Element(name);
                e.anchor   = anchor;
                e.dataType = dataType;
                e.merge    = merge;
                e.attrs    = attrs;
                e.items    = items;
                yield e;
            }
            case 0x02 -> new TextNode(b.str());
            case 0x03 -> {
                String dt  = b.str();
                String val = b.str();
                yield new ScalarNode(dt, coerce(dt, val));
            }
            case 0x04 -> new CommentNode(b.str());
            case 0x05 -> new RawTextNode(b.str());
            case 0x06 -> new EntityRefNode(b.str());
            case 0x07 -> new AliasNode(b.str());
            case 0x08 -> {
                String target = b.str();
                String data   = b.optStr();
                yield new PINode(target, data);
            }
            case 0x09 -> {
                String version    = b.str();
                String encoding   = b.optStr();
                String standalone = b.optStr();
                yield new XMLDeclNode(version, encoding, standalone);
            }
            case 0x0A -> {
                int nAttrs = b.u16();
                List<Attr> attrs = new ArrayList<>(nAttrs);
                for (int i = 0; i < nAttrs; i++) attrs.add(readAttr(b));
                yield new CXDirectiveNode(attrs);
            }
            case 0x0C -> {
                int        nItems = b.u16();
                List<Node> items  = new ArrayList<>(nItems);
                for (int i = 0; i < nItems; i++) items.add(readNode(b));
                yield new BlockContentNode(items);
            }
            // 0xFF = skip/DTD, no payload
            default -> new TextNode("");
        };
    }

    // ── Public: decode AST ────────────────────────────────────────────────────

    /**
     * Decode a binary AST payload (the bytes after the 4-byte length prefix)
     * into a {@link CXDocument}.
     *
     * Format:
     *   u8:  version (=1)
     *   u16: prolog_count  + prolog nodes
     *   u16: element_count + element nodes
     */
    public static CXDocument decodeAST(byte[] data) {
        BufReader b = new BufReader(data);
        /* int version = */ b.u8();  // currently unused

        int nProlog = b.u16();
        List<Node> prolog = new ArrayList<>(nProlog);
        for (int i = 0; i < nProlog; i++) prolog.add(readNode(b));

        int nElements = b.u16();
        List<Node> elements = new ArrayList<>(nElements);
        for (int i = 0; i < nElements; i++) elements.add(readNode(b));

        CXDocument doc = new CXDocument();
        doc.prolog   = prolog;
        doc.elements = elements;
        return doc;
    }

    // ── Public: decode Events ─────────────────────────────────────────────────

    /**
     * Decode a binary events payload (the bytes after the 4-byte length prefix)
     * into a list of {@link StreamEvent}s.
     *
     * Format:
     *   u32: event_count
     *   For each event:
     *     u8: type_id  (see binary.py / task spec for mapping)
     */
    public static List<StreamEvent> decodeEvents(byte[] data) {
        BufReader b = new BufReader(data);
        long n = b.u32();
        List<StreamEvent> events = new ArrayList<>((int) n);

        for (long i = 0; i < n; i++) {
            int tid = b.u8();
            StreamEvent e = switch (tid) {
                case 0x01 -> new StreamEvent("StartDoc");
                case 0x02 -> new StreamEvent("EndDoc");
                case 0x03 -> {
                    StreamEvent se = new StreamEvent("StartElement");
                    se.name     = b.str();
                    se.anchor   = b.optStr();
                    se.dataType = b.optStr();
                    se.merge    = b.optStr();
                    int nAttrs  = b.u16();
                    se.attrs    = new ArrayList<>(nAttrs);
                    for (int j = 0; j < nAttrs; j++) {
                        String aName    = b.str();
                        String aValStr  = b.str();
                        String aType    = b.str();
                        Object aVal     = coerce(aType, aValStr);
                        String aDt      = "string".equals(aType) ? null : aType;
                        se.attrs.add(new Attr(aName, aVal, aDt));
                    }
                    yield se;
                }
                case 0x04 -> {
                    StreamEvent se = new StreamEvent("EndElement");
                    se.name = b.str();
                    yield se;
                }
                case 0x05 -> {
                    StreamEvent se = new StreamEvent("Text");
                    se.value = b.str();
                    yield se;
                }
                case 0x06 -> {
                    StreamEvent se = new StreamEvent("Scalar");
                    String dt  = b.str();
                    se.dataType = dt;
                    se.value    = coerce(dt, b.str());
                    yield se;
                }
                case 0x07 -> {
                    StreamEvent se = new StreamEvent("Comment");
                    se.value = b.str();
                    yield se;
                }
                case 0x08 -> {
                    StreamEvent se = new StreamEvent("PI");
                    se.target = b.str();
                    se.data   = b.optStr();
                    yield se;
                }
                case 0x09 -> {
                    StreamEvent se = new StreamEvent("EntityRef");
                    se.value = b.str();
                    yield se;
                }
                case 0x0A -> {
                    StreamEvent se = new StreamEvent("RawText");
                    se.value = b.str();
                    yield se;
                }
                case 0x0B -> {
                    StreamEvent se = new StreamEvent("Alias");
                    se.value = b.str();
                    yield se;
                }
                default -> new StreamEvent("Unknown");
            };
            events.add(e);
        }
        return events;
    }
}

File: lang/kotlin/cxlib/src/main/kotlin/cx/BinaryDecoder.kt
package cx

import java.nio.ByteBuffer
import java.nio.ByteOrder

// ── StreamEvent ───────────────────────────────────────────────────────────────

data class StreamEvent(
    val type: String,
    val name: String? = null,
    val anchor: String? = null,
    val dataType: String? = null,
    val merge: String? = null,
    val attrs: List<Attr> = emptyList(),
    val value: Any? = null,
    val target: String? = null,
    val data: String? = null,
)

// ── BufReader ─────────────────────────────────────────────────────────────────

private class BufReader(data: ByteArray) {
    private val buf: ByteBuffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)

    fun u8(): Int = buf.get().toInt() and 0xFF

    fun u16(): Int = buf.short.toInt() and 0xFFFF

    fun u32(): Int = buf.int   // still LE; we treat as unsigned via toLong() where needed

    fun str(): String {
        val len = u32()
        val bytes = ByteArray(len)
        buf.get(bytes)
        return String(bytes, Charsets.UTF_8)
    }

    fun optStr(): String? {
        val flag = u8()
        return if (flag == 0) null else str()
    }
}

// ── scalar coercion ───────────────────────────────────────────────────────────

private fun coerce(typeStr: String, valueStr: String): Any? = when (typeStr) {
    "int"   -> valueStr.toLong()
    "float" -> valueStr.toDouble()
    "bool"  -> valueStr == "true"
    "null"  -> null
    else    -> valueStr
}

// ── BinaryDecoder ─────────────────────────────────────────────────────────────

object BinaryDecoder {

    // ── AST ───────────────────────────────────────────────────────────────────

    fun decodeAST(data: ByteArray): CXDocument {
        val buf = BufReader(data)
        @Suppress("UNUSED_VARIABLE")
        val version = buf.u8()          // currently always 1
        val prologCount = buf.u16()
        val prolog = (0 until prologCount).map { readNode(buf) }.toMutableList()
        val elemCount = buf.u16()
        val elements = (0 until elemCount).map { readNode(buf) }.toMutableList()
        return CXDocument(elements = elements, prolog = prolog)
    }

    private fun readAttr(buf: BufReader): Attr {
        val name     = buf.str()
        val valueStr = buf.str()
        val typeStr  = buf.str()
        val dt = if (typeStr == "string") null else typeStr
        return Attr(name = name, value = coerce(typeStr, valueStr), dataType = dt)
    }

    private fun readNode(buf: BufReader): Node {
        return when (val tid = buf.u8()) {
            0x01 -> {
                val name      = buf.str()
                val anchor    = buf.optStr()
                val dataType  = buf.optStr()
                val merge     = buf.optStr()
                val attrCount = buf.u16()
                val attrs     = (0 until attrCount).map { readAttr(buf) }.toMutableList()
                val childCount = buf.u16()
                val items     = (0 until childCount).map { readNode(buf) }.toMutableList()
                Element(name = name, anchor = anchor, merge = merge, dataType = dataType,
                        attrs = attrs, items = items)
            }
            0x02 -> TextNode(buf.str())
            0x03 -> {
                val typeStr  = buf.str()
                val valueStr = buf.str()
                ScalarNode(dataType = typeStr, value = coerce(typeStr, valueStr))
            }
            0x04 -> CommentNode(buf.str())
            0x05 -> RawTextNode(buf.str())
            0x06 -> EntityRefNode(buf.str())
            0x07 -> AliasNode(buf.str())
            0x08 -> {
                val target = buf.str()
                val data   = buf.optStr()
                PINode(target = target, data = data)
            }
            0x09 -> {
                val version    = buf.str()
                val encoding   = buf.optStr()
                val standalone = buf.optStr()
                XMLDeclNode(version = version, encoding = encoding, standalone = standalone)
            }
            0x0A -> {
                val attrCount = buf.u16()
                val attrs = (0 until attrCount).map { readAttr(buf) }
                CXDirectiveNode(attrs = attrs)
            }
            0x0C -> {
                val childCount = buf.u16()
                val items = (0 until childCount).map { readNode(buf) }
                BlockContentNode(items = items)
            }
            0xFF -> TextNode("")   // skip marker — no payload
            else  -> TextNode("")  // unknown type — no payload assumed
        }
    }

    // ── Events ────────────────────────────────────────────────────────────────

    fun decodeEvents(data: ByteArray): List<StreamEvent> {
        val buf = BufReader(data)
        val count = buf.u32()   // event_count is u32
        val events = ArrayList<StreamEvent>(count)
        repeat(count) {
            val tid = buf.u8()
            val evt = when (tid) {
                0x01 -> StreamEvent(type = "StartDoc")
                0x02 -> StreamEvent(type = "EndDoc")
                0x03 -> {
                    val name     = buf.str()
                    val anchor   = buf.optStr()
                    val dataType = buf.optStr()
                    val merge    = buf.optStr()
                    val attrCount = buf.u16()
                    val attrs = (0 until attrCount).map {
                        val n  = buf.str()
                        val vs = buf.str()
                        val t  = buf.str()
                        val dt = if (t == "string") null else t
                        Attr(name = n, value = coerce(t, vs), dataType = dt)
                    }
                    StreamEvent(type = "StartElement", name = name, anchor = anchor,
                                dataType = dataType, merge = merge, attrs = attrs)
                }
                0x04 -> StreamEvent(type = "EndElement",  name = buf.str())
                0x05 -> StreamEvent(type = "Text",        value = buf.str())
                0x06 -> {
                    val typeStr  = buf.str()
                    val valueStr = buf.str()
                    StreamEvent(type = "Scalar", dataType = typeStr,
                                value = coerce(typeStr, valueStr))
                }
                0x07 -> StreamEvent(type = "Comment",   value = buf.str())
                0x08 -> {
                    val target = buf.str()
                    val data   = buf.optStr()
                    StreamEvent(type = "PI", target = target, data = data)
                }
                0x09 -> StreamEvent(type = "EntityRef", value = buf.str())
                0x0A -> StreamEvent(type = "RawText",   value = buf.str())
                0x0B -> StreamEvent(type = "Alias",     value = buf.str())
                else  -> StreamEvent(type = "Unknown")
            }
            events.add(evt)
        }
        return events
    }
}
